docs/CSL-1925.md

```diff
@@ -0,0 +1,256 @@
+of it being generated randomly. `DeterministicMode` is there in order to be able to recover 
+the address at a later date.
+
+The wallet uses a function called `genUniqueAddress` to generate addresses which in turn
+uses the function shown below named `generateUnique`. In RandomMode this function tries 
+to ensure that the index used is unique according to an arbitrary predicate `isDuplicate`. 
+For addresses this predicate is called `doesWAddressExist` which checks storage for a 
+pre-existing address.
+
+In terms of BIP32 paths this extra index forms the `bH` in the derivation path `m/aH/bH` 
+where `bH` is the random index and H is there to denote that the function also runs 
+`isHardened` on the random index.
+
+## Random Address Generation
+
+Currently the random address generation / deterministic generation is coded as follows.
+
+```haskell
+generateUnique
+    :: (MonadIO m, MonadThrow m)
+    => Text
+    -> GenerationMode             -- ^ The type of address generation to use.
+    -> (Word32 -> m b)
+    -> (b -> m Bool)              -- ^ Whether this account is a duplicate of existing.
+    -> m b
+generateUnique desc RandomMode generator isDuplicate = loop (100 :: Int)
+  where
+    loop 0 = throwM . RequestError $
+             sformat (build%": generation of unique item seems too difficult, \
+                      \you are approaching the limit") desc
+    loop i = do
+        rand <- liftIO randomIO
+        value <- generator rand
+        bad <- orM
+            [ isDuplicate value
+            , pure $ isHardened rand      -- using hardened keys only for now
+            ]
+        if bad then
+            loop (i - 1)
+        else
+            return value
+generateUnique desc (DeterministicMode s1) generator notFit = do
+    value <- generator (fromIntegral s1)
+    whenM (notFit value) $
+        throwM . InternalError $
+        sformat (build%": this index is already taken")
+        desc
+    return value
+```
+
+The `genUniqueAddress` function attempts to prevent collisions by checking a local acid-state
+cache for the same address and retrying in case of failure. The function relies on 
+local storage instead of natural cryptographic constraints.
+
+Should the address being generated not be within storage or should storage fail, there is 
+then a non-negligible chance of collision which can occur and go unnoticed by dependent
+applications.
+
+Lastly `genUniqueAddress` is broken in the above implementation due to isHardened being 
+used inapropriately and being able to simplify the implementation via randomRIO. This will be 
+fixed as a matter of priority as a short term solution to the problem.
+
+# Non-Random Generation
+
+There is less chance of a collision in random generation when generating two
+distinct accounts on the same root key since the index is picked randomly. 
+In practice however this only matters when storage is not available and 
```

Comment:
That's an excellent point, and I would like us to spend a bit of time thinking more deeply on the implications of what would happen if the following happens:

- I am a large exchange with lots of addresses and a fully synced node, which means `acid-state` is sheltering me from collisions;

- I decide to delete my wallet, nuke my local storage and migrate everything to a new machine, due to server maintenance;

- I start the recovery process on the wallet, but I am still not fully synced.

- I call `POST /api/addresses` to get a branch new address for a registered-user

How would the backend behave? Will it reject the request or allow me to do that, with potentially disastrous consequences?

Reply:
I would assume perhaps naively that the fourth step here would wait until the sync in step 3 happens atomically. 

If the whole wallet is nuked and the recovery process has not been started then I would assume what would happen is:
1. A new root key is generated for the wallet based on the passphrase of the registered user.
2. This new root key is used to derive the first branch (either at random or sequentially).
3. The key resulting from the first derivation is used to derive the second branch (either at random or sequentially).

So I think actually in the case where the recovery process is atomic there is no problem. In the case where recovery fails altogether, the data may later need to be merged with the recoverable part of the failed data set.

Neither seem disastrous to me and I think that in this scenario it doesn't make a difference if you pick the index randomly or sequentially. I am not sure if I understood correctly your cases here though. :P